protocol/message_persistence.go

```diff
@@ -1305,26 +1305,50 @@ func (db sqlitePersistence) MessageByChatIDs(chatIDs []string, currCursor string
 	return result, newCursor, nil
 }
 
-func (db sqlitePersistence) OldestMessageWhisperTimestampByChatID(chatID string) (timestamp uint64, hasAnyMessage bool, err error) {
-	var whisperTimestamp uint64
-	err = db.db.QueryRow(
-		`
-			SELECT
-				whisper_timestamp
-			FROM
-				user_messages m1
-			WHERE
-				m1.local_chat_id = ?
-			ORDER BY substr('0000000000000000000000000000000000000000000000000000000000000000' || m1.clock_value, -64, 64) || m1.id ASC
-			LIMIT 1
-		`, chatID).Scan(&whisperTimestamp)
-	if err == sql.ErrNoRows {
-		return 0, false, nil
+func (db sqlitePersistence) OldestMessageWhisperTimestampByChatIDs(chatIDs []string) (map[string]uint64, error) {
+	if len(chatIDs) == 0 {
+		return nil, nil
 	}
+
+	args := make([]interface{}, len(chatIDs))
+	for i, id := range chatIDs {
+		args[i] = id
+	}
+
+	inVector := strings.Repeat("?, ", len(chatIDs)-1) + "?"
+	//nolint:gosec
+	query := fmt.Sprintf(`
+		SELECT
+			m1.local_chat_id,
+			m1.whisper_timestamp
+		FROM (
+			SELECT
+				m1.local_chat_id,
+				m1.whisper_timestamp,
+				ROW_NUMBER() OVER (PARTITION BY m1.local_chat_id ORDER BY substr('0000000000000000000000000000000000000000000000000000000000000000' || m1.clock_value, -64, 64) || m1.id ASC) AS rn
```

Comment:
I believe this is the first query in status-go using `partition by`. Not only for that, but I wonder if this query performs well when there are lots of rows in `user_messages`.

By any chance, did you run a query plan analysis on this query? One idea is to generate tons of rows (1k, 10k, 100k, and 1M) from a unit test and grab the test sqlite file to analyse.

Reply:
Lovely script ;) Seems pretty fast, although it's also running in your powerful dev machine.

> we had the index(idx_search_by_local_chat_id_sort_on_cursor) already, so I think the performance should be ok.

If you didn't see anything suspicious in the output of the query plan then it's great!

Reply:
Query plan looks straightforward indeed. Nice!